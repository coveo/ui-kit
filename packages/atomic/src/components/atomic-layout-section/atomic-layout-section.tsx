import {Component, Prop} from '@stencil/core';
import {Section} from './sections';

@Component({
  tag: 'atomic-layout-section',
  shadow: false,
})
export class AtomicLayoutSection {
  @Prop() public section!: Section;
  @Prop() public minWidth?: string;
  @Prop() public maxWidth?: string;
}
