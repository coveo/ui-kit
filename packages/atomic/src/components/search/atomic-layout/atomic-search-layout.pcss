.atomic-modal-opened {
  overflow-y: hidden;
}

atomic-layout-section[section='search'] {
  grid-area: atomic-section-search;
}

atomic-layout-section[section='facets'] {
  grid-area: atomic-section-facets;
}

atomic-layout-section[section='main'] {
  grid-area: atomic-section-main;
}

atomic-layout-section[section='status'] {
  grid-area: atomic-section-status;
}

atomic-layout-section[section='pagination'] {
  grid-area: atomic-section-pagination;
}

atomic-search-layout {
  width: 100%;
  height: 100%;
  display: none;
  grid-template-areas:
    '. atomic-section-search .'
    '. atomic-section-main .';
  grid-template-columns: var(--atomic-layout-spacing-x) minmax(0, 1fr) var(--atomic-layout-spacing-x);

  atomic-layout-section[section='search'] {
    margin: var(--atomic-layout-spacing-y) 0;
  }

  atomic-layout-section[section='facets'] {
    display: none;

    & * {
      margin-bottom: var(--atomic-layout-spacing-y);
    }
  }

  atomic-layout-section[section='main'] {
    margin-bottom: var(--atomic-layout-spacing-y);
  }

  atomic-layout-section[section='horizontal-facets'] {
    display: flex;
    flex-wrap: wrap;
    margin-bottom: var(--atomic-layout-spacing-y);
    gap: 0.5rem;

    & > atomic-segmented-facet-scrollable,
    & > atomic-segmented-facet,
    & > atomic-popover {
      max-width: 100%;
    }

    & > atomic-popover {
      display: none;
    }
  }

  atomic-layout-section[section='status'] {
    display: grid;
    justify-content: space-between;
    grid-template-areas:
      'atomic-breadbox      atomic-breadbox'
      'atomic-query-summary atomic-sort'
      'atomic-did-you-mean  atomic-did-you-mean';

    & > * {
      margin-bottom: var(--atomic-layout-spacing-y);
    }

    atomic-breadbox {
      grid-area: atomic-breadbox;
    }

    atomic-query-summary {
      grid-area: atomic-query-summary;
      align-self: center;
      overflow: hidden;
    }

    atomic-sort-dropdown {
      grid-area: atomic-sort;
      justify-self: end;
    }

    atomic-refine-toggle {
      grid-area: atomic-sort;
    }

    atomic-did-you-mean {
      grid-area: atomic-did-you-mean;
    }
  }

  atomic-layout-section[section='results'] {
    atomic-smart-snippet {
      margin-bottom: 1.5rem;
    }

    atomic-smart-snippet-suggestions {
      margin-bottom: 1.5rem;
    }
  }

  atomic-layout-section[section='pagination'] {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;

    atomic-load-more-results {
      width: 100%;
    }

    & > * {
      margin-top: var(--atomic-layout-spacing-y);
    }

    /* Approx. width of pager & results per page */
    @media only screen and (min-width: 50rem) {
      flex-direction: row;
    }
  }
}
