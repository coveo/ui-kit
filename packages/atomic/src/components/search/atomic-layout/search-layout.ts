import {
  findSection,
  sectionSelector,
} from '../../common/atomic-layout-section/sections';

export function makeDesktopQuery(mobileBreakpoint: string) {
  return `only screen and (min-width: ${mobileBreakpoint})`;
}
export function buildSearchLayout(
  element: HTMLElement,
  mobileBreakpoint: string
) {
  const id = element.id;
  const layoutSelector = `atomic-search-layout#${id}`;
  const cleanStatusSelector =
    'atomic-search-interface:not(.atomic-search-interface-no-results, .atomic-search-interface-error)';
  const mediaQuerySelector = `@media ${makeDesktopQuery(mobileBreakpoint)}`;

  const display = `${layoutSelector} { display: grid }`;
  const search = `${mediaQuerySelector} {
    ${layoutSelector} ${sectionSelector('search')} {
      justify-self: start;
      width: 80%;
    }
  }`;

  const facets = () => {
    const facetsSection = findSection(element, 'facets');
    const mainSection = findSection(element, 'main');
    if (!facetsSection || !mainSection) {
      return '';
    }

    const facetsMin = facetsSection.minWidth || '17rem';
    const facetsMax = facetsSection.maxWidth || '22rem';
    const mainMin = mainSection.minWidth || '50%';
    const mainMax = mainSection.maxWidth || '70rem';

    return `${mediaQuerySelector} {
      ${layoutSelector} {
        grid-template-areas:
        '. .                     atomic-section-search .'
        '. atomic-section-main   atomic-section-main   .';
        grid-template-columns:
          1fr minmax(${facetsMin}, ${facetsMax}) minmax(${mainMin}, ${mainMax}) 1fr;
        column-gap: var(--atomic-layout-spacing-x);
      }

      ${cleanStatusSelector} ${layoutSelector} {
        grid-template-areas:
          '. .                     atomic-section-search .'
          '. atomic-section-facets atomic-section-main   .'
          '. atomic-section-facets .                     .';
      }

      ${cleanStatusSelector} ${layoutSelector} ${sectionSelector('facets')} {
        display: block;
      }
    }`;
  };

  const refine = () => {
    const statusSection = findSection(element, 'status');
    if (!statusSection) {
      return '';
    }

    const refineToggle = statusSection.querySelector('atomic-refine-toggle');
    if (!refineToggle) {
      return '';
    }

    const statusSelector = `${layoutSelector} ${sectionSelector('status')}`;
    return `${statusSelector} atomic-sort-dropdown {
      display: none;
    }

    ${mediaQuerySelector} {
     ${statusSelector} atomic-sort-dropdown {
       display: block;
      }

      ${statusSelector} atomic-refine-toggle {
        display: none;
       }
    }`;
  };

  const horizontalFacets = () => {
    return `${mediaQuerySelector} {
      ${layoutSelector} ${sectionSelector(
        'horizontal-facets'
      )} > atomic-popover:not(.atomic-hidden) {
        display: block;
      }
    }`;
  };

  return [display, search, facets(), refine(), horizontalFacets()]
    .filter((declaration) => declaration !== '')
    .join('\n\n');
}
