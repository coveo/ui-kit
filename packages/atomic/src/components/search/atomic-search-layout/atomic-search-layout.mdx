import {Canvas, ArgTypes, Meta, Title, Description, Stories, Subtitle} from '@storybook/blocks';
import * as AtomicSearchLayoutStories from './atomic-search-layout.new.stories';
import {CanvasWithGithub} from '../../../../storybook-utils/documentation/canvas-with-github';

<Meta of={ AtomicSearchLayoutStories } />

<Title />
<Subtitle><Description/></Subtitle>

<CanvasWithGithub
  of={ AtomicSearchLayoutStories.Default }
  githubPath="TODO/atomic-search-layout/atomic-search-layout.ts"
/>

## Usage
TODO: Add usage instructions here.

## Examples
<Stories title={""}/>

## Reference
<ArgTypes />
