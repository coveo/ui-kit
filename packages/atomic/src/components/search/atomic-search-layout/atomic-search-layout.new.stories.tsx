import type {Meta, StoryObj as Story} from '@storybook/web-components';
import {parameters} from '@/storybook-utils/common/common-meta-parameters';
import {renderComponent} from '@/storybook-utils/common/render-component';
import {wrapInSearchInterface} from '@/storybook-utils/search/search-interface-wrapper';

const {decorator, play} = wrapInSearchInterface();

const meta: Meta = {
  component: 'atomic-search-layout',
  title: 'Search/Search Layout',
  id: 'atomic-search-layout',
  render: renderComponent,
  decorators: [decorator],
  parameters,
  play,
  args: {
    'slots-default': `<span>Layout content</span>`,
  },
};

export default meta;

export const Default: Story = {};
