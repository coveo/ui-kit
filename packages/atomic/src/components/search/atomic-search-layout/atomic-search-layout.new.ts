import {errorGuard} from '@/src/decorators/error-guard';
import {SearchLayoutMixin} from '@/src/mixins/search-layout-mixin';
import {LitElement, html, unsafeCSS} from 'lit';
import {customElement, property, state} from 'lit/decorators.js';
import {DEFAULT_MOBILE_BREAKPOINT} from '../../../utils/replace-breakpoint';
import styles from './atomic-search-layout.tw.css';

/**
 * The `atomic-search-layout` helps organize elements in the page.
 */
@customElement('atomic-search-layout')
export class AtomicSearchLayout extends SearchLayoutMixin(
  LitElement,
  unsafeCSS(styles)
) {
  @state() error!: Error;

  /**
   * CSS value that defines where the layout goes from mobile to desktop.
   * e.g., 800px, 65rem.
   */
  @property({type: String, reflect: true, attribute: 'mobile-breakpoint'})
  mobileBreakpoint: string = DEFAULT_MOBILE_BREAKPOINT;

  createRenderRoot() {
    return this;
  }

  @errorGuard()
  render() {
    return html`<slot></slot>`;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'atomic-search-layout': AtomicSearchLayout;
  }
}
