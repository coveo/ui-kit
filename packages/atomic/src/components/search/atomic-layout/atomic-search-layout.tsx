import {Component, Element, Prop} from '@stencil/core';
import {
  InitializableComponent,
  InitializeBindings,
} from '../../../utils/initialization-utils';
import {DEFAULT_MOBILE_BREAKPOINT} from '../../../utils/replace-breakpoint';
import {randomID} from '../../../utils/utils';
import {Bindings} from '../atomic-search-interface/atomic-search-interface';
import {buildSearchLayout} from './search-layout';

/**
 * The `atomic-search-layout` helps organize elements in the page.
 */
@Component({
  tag: 'atomic-search-layout',
  styleUrl: 'atomic-search-layout.pcss',
  shadow: false,
})
export class AtomicSearchLayout implements InitializableComponent<Bindings> {
  public error!: Error;

  @Element() private host!: HTMLElement;
  /**
   * CSS value that defines where the layout goes from mobile to desktop.
   * e.g., 800px, 65rem.
   */
  @Prop({reflect: true}) public mobileBreakpoint = DEFAULT_MOBILE_BREAKPOINT;
  @InitializeBindings() public bindings!: Bindings;

  public componentDidLoad() {
    const id = this.host.id || randomID('atomic-search-layout-');
    this.host.id = id;
    const styleTag = this.bindings.createStyleElement();
    styleTag.innerHTML = buildSearchLayout(this.host, this.mobileBreakpoint);
    this.host.appendChild(styleTag);
  }
}
