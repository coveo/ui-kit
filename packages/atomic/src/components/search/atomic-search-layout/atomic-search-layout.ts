import {LitElement} from 'lit';
import {customElement, property, state} from 'lit/decorators.js';
import {LayoutStylesController} from '@/src/components/common/layout/layout-styles-controller';
import {watch} from '@/src/decorators/watch';
import {ChildrenUpdateCompleteMixin} from '@/src/mixins/children-update-complete-mixin';
import {LightDomMixin} from '@/src/mixins/light-dom';
import {DEFAULT_MOBILE_BREAKPOINT} from '@/src/utils/replace-breakpoint-utils';
import styles from './atomic-search-layout.tw.css';
import {buildSearchLayout} from './search-layout';

/**
 * The `atomic-search-layout` helps organize elements in the page.
 *
 * @cssprop --atomic-layout-max-search-box-input-width: The maximum width that the search box input will take.
 * @cssprop --atomic-layout-max-search-box-double-suggestions-width: The maximum width that the search box suggestions will take when displaying a double list.
 * @cssprop --atomic-layout-search-box-left-suggestions-width: The width of the left list when displaying a double list.
 */
@customElement('atomic-search-layout')
export class AtomicSearchLayout extends LightDomMixin(
  ChildrenUpdateCompleteMixin(LitElement)
) {
  static styles = [styles];

  @state() error!: Error;

  /**
   * The viewport width at which the layout goes from desktop to mobile.
   * E.g., 800px, 65rem.
   */
  @property({type: String, reflect: true, attribute: 'mobile-breakpoint'})
  mobileBreakpoint: string = DEFAULT_MOBILE_BREAKPOINT;

  private layoutStylesController = new LayoutStylesController(
    this,
    buildSearchLayout,
    'atomic-search-layout-'
  );

  connectedCallback() {
    super.connectedCallback();
    this.emitBreakpointChangeEvent();
  }

  private emitBreakpointChangeEvent() {
    this.dispatchEvent(
      new CustomEvent('atomic-layout-breakpoint-change', {
        detail: {breakpoint: this.mobileBreakpoint},
        bubbles: true,
        composed: true,
      })
    );
  }

  @watch('mobileBreakpoint')
  public onMobileBreakpointChange() {
    this.layoutStylesController.updateStyles();
    this.emitBreakpointChangeEvent();
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'atomic-search-layout': AtomicSearchLayout;
  }
}
