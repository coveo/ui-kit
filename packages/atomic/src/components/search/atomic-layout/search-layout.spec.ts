import {newSpecPage} from '@stencil/core/testing';
import {AtomicLayoutSection} from '../../common/atomic-layout-section/atomic-layout-section';
import {AtomicSearchLayout} from './atomic-search-layout';
import {buildSearchLayout} from './search-layout';

describe('Search Layout', () => {
  describe('#buildSearchLayout', () => {
    async function getSearchLayout(html: string) {
      const page = await newSpecPage({
        components: [AtomicLayoutSection, AtomicSearchLayout],
        html,
      });
      return buildSearchLayout(
        page.body.querySelector('atomic-search-layout')!,
        '1024px'
      );
    }

    it(`when the atomic-search-layout component is empty
    should match snapshot`, async () => {
      const result = await getSearchLayout(
        '<atomic-search-layout id="abc"></atomic-search-layout>'
      );
      expect(result).toMatchSnapshot();
    });

    it(`when the atomic-search-layout component contains a main & facets sections
    should match snapshot`, async () => {
      const result = await getSearchLayout(`<atomic-search-layout id="abc">
            <atomic-layout-section section="facets"></atomic-layout-section>
            <atomic-layout-section section="main"></atomic-layout-section>
        </atomic-search-layout>`);
      expect(result).toMatchSnapshot();
    });

    it(`when the atomic-search-layout has custom breakpoint & min-width, max-width values
    should match snapshot`, async () => {
      const result =
        await getSearchLayout(`<atomic-search-layout id="abc" breakpoint="600px">
            <atomic-layout-section section="facets" min-width="250px" max-width="300px"></atomic-layout-section>
            <atomic-layout-section section="main" min-width="300px" max-width="1200px"></atomic-layout-section>
        </atomic-search-layout>`);
      expect(result).toMatchSnapshot();
    });

    it(`when the atomic-search-layout component contains a status section which contains a atomic-refine-toggle component
    should match snapshot`, async () => {
      const result = await getSearchLayout(`<atomic-search-layout id="abc">
            <atomic-layout-section section="status">
                <atomic-refine-toggle></atomic-refine-toggle>
            </atomic-layout-section>
        </atomic-search-layout>`);
      expect(result).toMatchSnapshot();
    });
  });
});
