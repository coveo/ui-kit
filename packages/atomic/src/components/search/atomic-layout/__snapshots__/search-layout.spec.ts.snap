// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Layout #buildSearchLayout when the atomic-search-layout component contains a main & facets sections
    should match snapshot 1`] = `
"atomic-search-layout#abc { display: grid }

@media only screen and (min-width: 1024px) {
    atomic-search-layout#abc atomic-layout-section[section="search"] {
      justify-self: start;
      width: 80%;
    }
  }

@media only screen and (min-width: 1024px) {
      atomic-search-layout#abc {
        grid-template-areas:
        '. .                     atomic-section-search .'
        '. atomic-section-main   atomic-section-main   .';
        grid-template-columns:
          1fr minmax(17rem, 22rem) minmax(50%, 70rem) 1fr;
        column-gap: var(--atomic-layout-spacing-x);
      }

      atomic-search-interface:not(.atomic-search-interface-no-results, .atomic-search-interface-error) atomic-search-layout#abc {
        grid-template-areas:
          '. .                     atomic-section-search .'
          '. atomic-section-facets atomic-section-main   .'
          '. atomic-section-facets .                     .';
      }

      atomic-search-interface:not(.atomic-search-interface-no-results, .atomic-search-interface-error) atomic-search-layout#abc atomic-layout-section[section="facets"] {
        display: block;
      }
    }

@media only screen and (min-width: 1024px) {
      atomic-search-layout#abc atomic-layout-section[section="horizontal-facets"] > atomic-popover:not(.atomic-hidden) {
        display: block;
      }
    }"
`;

exports[`Search Layout #buildSearchLayout when the atomic-search-layout component contains a status section which contains a atomic-refine-toggle component
    should match snapshot 1`] = `
"atomic-search-layout#abc { display: grid }

@media only screen and (min-width: 1024px) {
    atomic-search-layout#abc atomic-layout-section[section="search"] {
      justify-self: start;
      width: 80%;
    }
  }

atomic-search-layout#abc atomic-layout-section[section="status"] atomic-sort-dropdown {
      display: none;
    }

    @media only screen and (min-width: 1024px) {
     atomic-search-layout#abc atomic-layout-section[section="status"] atomic-sort-dropdown {
       display: block;
      }

      atomic-search-layout#abc atomic-layout-section[section="status"] atomic-refine-toggle {
        display: none;
       }
    }

@media only screen and (min-width: 1024px) {
      atomic-search-layout#abc atomic-layout-section[section="horizontal-facets"] > atomic-popover:not(.atomic-hidden) {
        display: block;
      }
    }"
`;

exports[`Search Layout #buildSearchLayout when the atomic-search-layout component is empty
    should match snapshot 1`] = `
"atomic-search-layout#abc { display: grid }

@media only screen and (min-width: 1024px) {
    atomic-search-layout#abc atomic-layout-section[section="search"] {
      justify-self: start;
      width: 80%;
    }
  }

@media only screen and (min-width: 1024px) {
      atomic-search-layout#abc atomic-layout-section[section="horizontal-facets"] > atomic-popover:not(.atomic-hidden) {
        display: block;
      }
    }"
`;

exports[`Search Layout #buildSearchLayout when the atomic-search-layout has custom breakpoint & min-width, max-width values
    should match snapshot 1`] = `
"atomic-search-layout#abc { display: grid }

@media only screen and (min-width: 1024px) {
    atomic-search-layout#abc atomic-layout-section[section="search"] {
      justify-self: start;
      width: 80%;
    }
  }

@media only screen and (min-width: 1024px) {
      atomic-search-layout#abc {
        grid-template-areas:
        '. .                     atomic-section-search .'
        '. atomic-section-main   atomic-section-main   .';
        grid-template-columns:
          1fr minmax(250px, 300px) minmax(300px, 1200px) 1fr;
        column-gap: var(--atomic-layout-spacing-x);
      }

      atomic-search-interface:not(.atomic-search-interface-no-results, .atomic-search-interface-error) atomic-search-layout#abc {
        grid-template-areas:
          '. .                     atomic-section-search .'
          '. atomic-section-facets atomic-section-main   .'
          '. atomic-section-facets .                     .';
      }

      atomic-search-interface:not(.atomic-search-interface-no-results, .atomic-search-interface-error) atomic-search-layout#abc atomic-layout-section[section="facets"] {
        display: block;
      }
    }

@media only screen and (min-width: 1024px) {
      atomic-search-layout#abc atomic-layout-section[section="horizontal-facets"] > atomic-popover:not(.atomic-hidden) {
        display: block;
      }
    }"
`;
